DESIGN.md

```diff
@@ -0,0 +1,147 @@
+00:00:00, 00:15:00, 00:30:00, or 00:45:00). For example, if a user creates a
+bucket at 06:15:01, the operation will have a timestamp of 06:15:00. All
+timestamps are then formatted as a UNIX epoch expressed in milliseconds. During
+a listing of metrics, then, we can know that this operation occurred sometime
+between 06:15:00 and 06:29:59.
+
+#### Example
+
+| current time | timestamp | UNIX epoch timestamp |
+|--------------|-----------|----------------------|
+| 06:15:01     | 06:15:00  | 1483280100000        |
+| 06:29:59     | 06:15:00  | 1483280100000        |
+| 06:31:00     | 06:30:00  | 1483281000000        |
+| 07:01:00     | 07:00:00  | 1483282800000        |
+
+![timestamp graphic](res/timestamp-graphic.png)
+
+### Data Storage
+
+Utapi uses Redis as a database for storage of its metrics values. Accordingly,
+it uses two different Redis data types: Sorted Sets and Strings. This document
+describes how these two data types are used by Utapi. For further information on
+data types see the Redis [documentation](https://redis.io/topics/data-types).
+
+#### Sorted Sets
+
+The Redis keys storing metrics for the number of objects and the storage
+utilized are recorded with a Sorted Set data type. We use this data type to
+associate the value of the key with its timestamp (in Redis terminology, a
+score). In this way, these two metrics hold *stateful* data. That is, the key's
+value represents the state of a metric at that point in history.
+
+With a Sorted Set, then, we can create a list of values ordered by their
+timestamp. This ordered nature is especially useful for Utapi during a listing
+of metrics, as it allows for speedy retrieval of the nearest starting and ending
+values.
+
+#### Strings
+
+The remaining Redis keys recorded by Utapi use a String data type. These include
+metrics for incoming bytes, outgoing bytes, and all S3 operations (e.g.,
+'CreateBucket', 'PutObject', etc.).
+
+Moreover, there are also global counters associated with metrics for the number
+of objects and the storage utilized. Such counters are updated during any
+applicable operation. For example, when uploading or deleting an object, the
+counter for the number of objects increments or decrements, respectively.
+
+#### Example
+
+Steps occurring during a 'PutObject' request:
+
+1. If the new object overwrites a pre-existing object, the counter for the
+    number of objects remains unchanged. Otherwise it increments by one.
+
+2. If the new object overwrites a pre-existing object, the counter for the
+    storage utilized increments by the difference between the byte size of the
+    object and byte size of the object being overwritten. Otherwise, it
+    increments by the byte size of the new object.
+
+3. The metric for the incoming bytes increments by the byte size of the new
+    object.
+
+4. The metric for the 'PutObject' operation increments by one.
+
+5. The Sorted Set keys are updated to the value of their respective counter.
```

Comment:
The Sorted Set keys **(storage used, number of objects)** are updated to the value...